design/open/design-datasets.md

```diff
@@ -0,0 +1,211 @@
+  metadata.
+
+## Analysis
+
+We currently support these 3 ways of defining tables:
+
+* Implicitly, by discovering that a given path holds a table.  This is our
+  current Delta Lake and Delta-diff support.
+* Externally, by deferring to an external catalog.  This is our metastore
+  support (HMS is the catalog) and our Iceberg support (SparkCatalog is the
+  catalog).
+* Internally, by effectively serving the catalog.  This is our lakeFS Delta
+  Sharing, except that the Delta Sharing protocol does not call this a
+  "catalog".
+
+### Implicit definitions
+
+Implicit definitions give users the _easiest_ path.  However they have
+multiple limitations:
+
+* Require discoverability of the table by examining only object metadata.
+  So supported only by "modern" table formats: Iceberg and Delta.  Formats
+  such as prefixes containing Parquet files _cannot_ support all operations:
+  - It is not possible to report table metadata efficiently.
+  - It may not be possible to list all tables.
+  - It will _not_ be possible to support named tables, all tables must be
+    defined by paths.
+
+Using implicit definitions necessarily leaks through the table abstraction.
+There is no useful general case here.  Users will find it easy to request
+functionality that uses implicit definitions.
+
+#### Possible future features
+
+Define an implicit "Parquet folder" as a folder that holds many `.parquet`
+files, or a folder that holds many `<key>=<value>` subfolders with some
+`.parquet` files under them.  Then we could add to the GUI a display of a
+sample DuckDB query when looking at a Parquet folder:
+
+### External definitions
+
+External definitions in catalogs may be the easiest way to give _some_ users
+a fully-functional experience, for _some_ catalogs.  Their principal
+limitations:
+
+* Risk of fragmentation: If there are many popular external catalogs we will
+  need to support many of them.  Different catalogs support different
+  use-cases, so we should expect to have to support multiple catalogs.
+* Limitations due to the external definition: Because existing catalog
+  implementations were designed without lakeFS in mind, we will not be able
+  to add lakeFS features to all of them.  For instance the HMS protocol is
+  Thrift-based and does not support refs-as-schemas.
+
+#### Possible future features
+
+Support more catalog types.  Each catalog adds implementation effort.
+
+Support catalog wrappers that work with multiple external catalogs.  For
+instance, support wrapping additional catalogs under the Iceberg catalog.
+
+### Internal definition
+
+An internal definition should allow us to support _all_ lakeFS features that
+systems can handle.  Because it reduces external dependencies, it is
+probably the fastest way for us to implement "table" features on top of
+lakeFS -- we have seen this in the Iceberg catalog wrapper which is barely
+more than an internal definition, and in the lakeFS Delta Sharing server.
```

Comment:
not sure that I understand it... How is the Iceberg catalog wrapper an internal definition?